pkg/registry/generic/registry/store.go

```diff
@@ -235,6 +240,49 @@ func (e *Store) Create(ctx api.Context, obj runtime.Object) (runtime.Object, err
 	return out, nil
 }
 
+// shouldDelete checks if a Update is removing all the object's finalizers. If so,
+// it further checks if the object's DeletionGracePeriodSeconds is 0. If so, it
+// returns true.
+func (e *Store) shouldDelete(ctx api.Context, key string, obj, existing runtime.Object) bool {
+	if !EnableGarbageCollector {
+		return false
+	}
+	newMeta, err := api.ObjectMetaFor(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return false
+	}
+	oldMeta, err := api.ObjectMetaFor(existing)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return false
+	}
+	return len(newMeta.Finalizers) == 0 && oldMeta.DeletionGracePeriodSeconds != nil && *oldMeta.DeletionGracePeriodSeconds == 0
```

Comment:
Your statement is correct. My reasoning is: If oldMeta.DeletionGracePeriodSeconds==0, and the object still exists, it could only be that the finalizers were not empty, or the apiserver had crashed, either way we should delete the object now.